### PR TITLE
Switch DATE type to be int32

### PR DESCRIPTION
**Pull request by apavlo**
The `DATE` type was originally set to be uint32_t. The problem with this is that we had set `PELOTON_DATE_MIN` to be the same as `PELOTON_DATE_NULL`, which means that you could never have null dates.

This PR switches the `DATE` to be a int32_t. I also added new tests because the coverage for this code was nearly zero.

It also includes SQL tests that check whether we check that you can't insert the min values for types and end up with nulls.

I also fixed Value::Min/Max for some types.

https://www.youtube.com/watch?v=1XV-dWStHOI

### CODE REVIEW

File: test/type/value_test.cpp [hunk 1]
```diff
@@ -59,6 +59,35 @@ TEST_F(ValueTests, HashTest) {
   }  // FOR
 }
 
+TEST_F(ValueTests, MinMaxTest) {
+  for (auto col_type : valueTestTypes) {
+    auto maxVal = type::Type::GetMaxValue(col_type);
```
Comment (phisiart) [inline test/type/value_test.cpp]: Same here, naming convention.

File: test/type/date_value_test.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+          EXPECT_EQ(expected, result == type::CMP_TRUE);
+          EXPECT_EQ(!expected, result == type::CMP_FALSE);
+        }
+      }
+    }
+  }
+}
+
+TEST_F(DateValueTests, NullToStringTest) {
+  auto valNull = type::ValueFactory::GetNullValueByType(type::TypeId::DATE);
```
[the remark this answers is not in the document]
Comment (apavlo) [reply test/type/date_value_test.cpp]: Fixed.

File: test/type/date_value_test.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  type::Value val0 =
+      type::ValueFactory::GetDateValue(static_cast<int32_t>(1000000));
+  type::Value val1 = val0.Copy();
+  EXPECT_EQ(val0.CompareEquals(val1) == type::CMP_TRUE, true);
+}
+
+TEST_F(DateValueTests, CastTest) {
+  type::Value result;
+
+  auto strNull = type::ValueFactory::GetNullValueByType(type::TypeId::VARCHAR);
```
[the remark this answers is not in the document]
Comment (apavlo) [reply test/type/date_value_test.cpp]: Fixed.

File: test/type/date_value_test.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+  EXPECT_EQ(result.GetTypeId(), valNull.GetTypeId());
+
+  result = valNull.CastAs(type::TypeId::VARCHAR);
+  EXPECT_TRUE(result.IsNull());
+  EXPECT_EQ(result.CompareEquals(strNull) == type::CMP_NULL, true);
+  EXPECT_EQ(result.GetTypeId(), strNull.GetTypeId());
+
+  EXPECT_THROW(valNull.CastAs(type::TypeId::BOOLEAN), peloton::Exception);
+
+  auto valValid =
```
[the remark this answers is not in the document]
Comment (apavlo) [reply test/type/date_value_test.cpp]: Fixed.

File: test/type/value_factory_test.cpp [hunk 5]
```diff
@@ -55,6 +61,13 @@ int64_t RANDOM64() {
   return 1;
 }
 
+TEST_F(ValueFactoryTests, ZeroValueTest) {
+  for (auto col_type : valueFactoryTestTypes) {
+    auto zeroVal = type::ValueFactory::GetZeroValueByType(col_type);
```
[the remark this answers is not in the document]
Comment (apavlo) [reply test/type/value_factory_test.cpp]: Fixed.

File: test/type/value_factory_test.cpp [hunk 6]
```diff
@@ -134,49 +147,28 @@ TEST_F(ValueFactoryTests, CastTest) {
...
-  EXPECT_EQ(v8.GetAs<int64_t>(), type::PELOTON_INT64_MIN);
-
-  type::Value v9 = (type::Value::DeserializeFrom(in, type::Type::GetInstance(type::TypeId::DECIMAL)->GetTypeId(), nullptr));
-  EXPECT_EQ(v9.GetTypeId(), type::TypeId::DECIMAL);
-  EXPECT_EQ(v9.GetAs<double>(), type::PELOTON_DECIMAL_MAX);
-  type::Value v10 = (type::Value::DeserializeFrom(in, type::Type::GetInstance(type::TypeId::DECIMAL)->GetTypeId(), nullptr));
-  EXPECT_EQ(v10.GetTypeId(), type::TypeId::DECIMAL);
-  EXPECT_EQ(v10.GetAs<double>(), type::PELOTON_DECIMAL_MIN);
+  for (auto col_type : valueFactoryTestTypes) {
+    for (int i = 0; i < 2; i++) {
```
[the remark this answers is not in the document]
Comment (apavlo) [reply test/type/value_factory_test.cpp]: Fixed.

File: test/type/value_test.cpp (hunk 1 above)
Comment (apavlo) [reply test/type/value_test.cpp]: Fixed.
